src/Database/Persist/URL.hs

```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Database.Persist.URL
+    ( fromDatabaseUrl
+    ) where
+
+import Control.Monad (MonadPlus, mzero)
+import Data.ByteString (ByteString, uncons)
+import Data.Monoid ((<>))
+import Database.Persist.Postgresql (PostgresConf(..))
+import URI.ByteString
+    ( Authority(..)
+    , Host(..)
+    , Port(..)
+    , URI(..)
+    , UserInfo(..)
+    , parseURI
+    , strictURIParserOptions
+    )
+
+import qualified Data.ByteString.Char8 as Char8
+
+-- | Build a @'PostgresConf'@ by parsing a database URL String
+fromDatabaseUrl :: MonadPlus m => Int -> ByteString -> m PostgresConf
```

Comment:
Re-reading this thread, I realize my comment here was sort of confusing without context:

After the initial `MonadPlus` vs `MonadThrow` discussion, we tried using `MonadFail` instead in order to preserve the ability to pull out an error message. We chose that over `MonadThrow` because it's going to be part of `base` in GHC 8, and Aeson's `Parser` had an instance of it. (Aeson still has no instance of `MonadThrow`).

It turns out that `MonadPlus` is the only one usable without requiring the latest version of Aeson, however. Maybe we'll go back to that, or maybe we'll give up on the idea of avoiding `Either`/`Maybe` in `Settings.hs`.